improver/nowcasting/utilities.py

```diff
@@ -103,6 +102,110 @@ def process(self, radar_data, coverage):
         return cube_list.merge_cube()
 
 
+class FillRadarHoles(BasePlugin):
+    """Fill in small "no data" regions in the radar composite by interpolating
+    in log rainrate space."""
+
+    @staticmethod
+    def _find_and_interpolate_speckle(cube, copy=False):
+        """Identify and interpolate "speckle" points, where "speckle" is defined
+        as areas of "no data" that are small enough to fill by interpolation
+        without affecting data integrity.  We would not wish to interpolate large
+        areas as this may give false confidence in "no precipitation", where in
+        fact precipitation exists in a "no data" region.
+
+        The constants for speckle identification ("p_masked" and "r_speckle")
+        have been empirically tuned for UK radar data.  With the constants
+        as set, this method will flag "holes" of up to 24 pixels in size.
+        The interpolation radius has been chosen to match these constants,
+        as the smallest radius that ensures there will always be valid data in
+        the neighbourhood (25 pixels) over which averaging is performed.
+        Masked pixels near the border of the input data array will not be
+        interpolated.
+
+        Args:
+            cube (iris.cube.Cube):
+                Cube containing rainrates (mm/h).
+            copy (bool):
+                If true, creates a copy of the cube and returns the copy, otherwise writes it in place.
+        Returns:
+            iris.cube.Cube:
+                Cube containing the interpolated rainrates (mm/h)."""
+        # proportion of masked neighbours below which a pixel is considered to
+        # be isolated "speckle", which can be filled in by interpolation
+        p_masked = 0.3
+        # radius of neighbourhood over which to search for masked neighbours
+        r_speckle = 4
+        # radius of neighbourhood from which to calculate interpolated values
+        r_interp = 2
+
+        min_rr_mmh = 0.001
+
+        if copy:
+            cube_new = cube.copy()
+        else:
+            cube_new = cube
+
+        window_shape = (2 * r_speckle) + 1
+        window_shape = (window_shape, window_shape)
+        max_mask_values = window_shape[0] * window_shape[1] * p_masked
+
+        mask_windows = neighbourhood_tools.pad_and_roll(
+            cube_new.data.mask, window_shape, mode="constant", constant_values=1
+        )
+        data_windows = neighbourhood_tools.pad_and_roll(
+            cube_new.data, window_shape, mode="constant", constant_values=np.nan
+        )
+
+        # Find where the center pixel is masked and the total masked in the bhood doesn't exceed the threshold
+        indices = np.where(
+            (mask_windows[..., r_speckle, r_speckle] == 1)
+            & (np.sum(mask_windows, axis=(-2, -1)) < max_mask_values)
+        )
+        data_windows[indices]
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5471664)